symengine/parser.cpp

```diff
@@ -0,0 +1,235 @@
+                        break;
+                    case '(':
+                        result = functionify(iter, expr);
+                        break;
+                    case ')':
+                        continue;
+                    case ',':
+                        continue;
+                }
+
+                result_set = true;
+                expr_is_symbol = false;
+            }
+            else
+            {
+                expr += s[iter];
+
+                int ascii = s[iter] - '0';
+                if (ascii < 0 or ascii > 9)
+                    expr_is_symbol = true;
+
+                if (iter == h-1)
+                {
+                    if (expr_is_symbol)
+                        result = symbol(expr);
+                    else
+                        result = integer(std::stoi(expr));
+                }
+            }
+        }
+
+        return result;
+    }
+
+    bool is_operator(int iter)
+    {
+        if (iter >= 0 and iter < (int)s_len)
+            if (OPERATORS.find(s[iter]) != OPERATORS.end())
+                return true;
+        return false;
+    }
+
+    RCP<const Basic> functionify(unsigned int& iter, std::string expr)
+    {
+        RCP<const Basic> param1 = parse_string(iter+1, operator_end[iter]);
+        iter = operator_end[iter] - 1;
+
+        if(expr == "") return param1;
+
+        if(expr == "sin") return sin(param1);
+        if(expr == "cos") return cos(param1);
+        if(expr == "tan") return tan(param1);
+        if(expr == "cot") return cot(param1);
+        if(expr == "csc") return csc(param1);
+        if(expr == "sec") return sec(param1);
+
+        if(expr == "arcsin") return asin(param1);
+        if(expr == "arccos") return acos(param1);
+        if(expr == "arctan") return atan(param1);
+        if(expr == "arcsec") return asec(param1);
+        if(expr == "arccsc") return acsc(param1);
+        if(expr == "arccot") return acot(param1);
+
+        if(expr == "sinh") return sinh(param1);
+        if(expr == "cosh") return cosh(param1);
+        if(expr == "tanh") return tanh(param1);
+        if(expr == "coth") return coth(param1);
+        // implement sech, csch ?
+
+        if(expr == "arcsinh") return asinh(param1);
+        if(expr == "arccosh") return acosh(param1);
+        if(expr == "arctanh") return atanh(param1);
+        if(expr == "arcsech") return asech(param1);
+        if(expr == "arccoth") return acoth(param1);
+        // implement acsch ?
+
+        if(expr == "gamma") return gamma(param1);
+        if(expr == "dirichlet_eta") return dirichlet_eta(param1);
+        if(expr == "lambertw") return lambertw(param1);
+        if(expr == "sqrt") return sqrt(param1);
+        if(expr == "abs") return abs(param1);
+        if(expr == "exprp") return exp(param1);
+        if(expr == "ln") return log(param1);
+
+        if(expr == "zeta") if(s[iter+1] != ',') return zeta(param1);
+        if(expr == "log") if(s[iter+1] != ',') return log(param1);
```

Comment:
I really don't think anything faster than linear is possible. I have to search if 'any' of them matches, and do a unique thing on a match. I am not looking to 'find' an element in a array/map. Whatever way it is implemented, it will have to boil down to checking each the `expr` against each of these terms. (or something equivalent to these linear checks)

One complicated way to optimize maybe to make a search tree based on the letters of the `expr`. But I feel that is an overkill. Maybe do it after the main tasks?